### PR TITLE
Fix some issues with static data in SIDE_MODULEs

### LINKED ISSUE #6310: Function pointers are wrong in the second loaded SIDE_MODULE
I'm building Numpy and Python for iodide-project/iodide, an in-browser scientific computing notebook.

Numpy has a number of C extensions, but only 2 are "required" for basic functionality: `multiarray.so` and `umath.so`.  I have a recipe that builds Python with `dlopen` support as a `MAIN_MODULE` and builds these two extension modules as `SIDE_MODULE`.  They import from Python fine.  The first module, `multiarray`, works correctly.  However, the second loaded module, `umath`, seems to have incorrect function pointer tables.  When calling anything involving a function pointer (which is most functions, because the library is basically designed around dynamic dispatch on the types of the arguments), the function that actually gets called seems "random", and will actually change on subsequent compiles.

Python plays fast-and-loose with function pointers, and basically requires the use of `EMULATE_FUNCTION_POINTER_CASTS`, which I think is related to my problem.  Due to #6309, I'm outputting asm.js and not setting `-s WASM=1`, if that matters here.

Any tips as to how to investigate the issue further?

### PR DESCRIPTION
**Pull request by mdboom**
This fixes #6310, though my hunches about the root cause turned out to be wrong.

There are two fixes here:

1) The static data in a module generally consists of content initialized with an array literal, followed by zeroinitialized content.  The latter was not getting zero-initialized when the static chunk is allocated.  Changing the `malloc` within `getMemory` to `calloc` seems to resolve this.

2) Static memory chunks in side modules were not getting aligned correctly.  This seems to be due to a parentheses placement bug.  This seems to manifest when there are short (empty) strings in the last 8 bytes at the end of the zero-initialized area.

### CODE REVIEW

File: src/jsifier.js [hunk 1]
```diff
@@ -432,7 +432,7 @@ function JSify(data, functionsOnly) {
           print('STATIC_BASE = GLOBAL_BASE;\n');
           print('STATICTOP = STATIC_BASE + ' + Runtime.alignMemory(Variables.nextIndexedOffset) + ';\n');
         } else {
-          print('gb = alignMemory(getMemory({{{ STATIC_BUMP }}}, ' + MAX_GLOBAL_ALIGN + ' || 1));\n');
+          print('gb = alignMemory(getMemory({{{ STATIC_BUMP }}}), ' + MAX_GLOBAL_ALIGN + ' || 1);\n');
```
Comment (kripken) [inline src/jsifier.js]: Oh wow, the bug was the wrong placement of parens? Good find!

We should figure out how to add a test for this sensitive case.

File: src/preamble.js [hunk 2]
```diff
@@ -361,7 +361,7 @@ function allocate(slab, types, allocator, ptr) {
 function getMemory(size) {
   if (!staticSealed) return staticAlloc(size);
   if (!runtimeInitialized) return dynamicAlloc(size);
-  return _malloc(size);
+  return _calloc(1, size);
```
Comment (kripken) [inline src/preamble.js]: A downside to doing this is that it means getMemory is slower for all other users. I think we should instead find the place calling getMemory here and zero out the memory afterwards.

Another issue is the use of calloc - we can't depend on it always existing (unlike malloc which we do link in by default). We could ensure it is linked in when necessary, but zeroing out the memory afterwards is probably simpler and good enough here.

Comment (mdboom) [reply src/preamble.js]: Sure.  The place that matters is where the static memory for a SIDE_MODULE is allocated -- the other part of this patch in `jsifier.js`.  I'll update this to fix it there.

Comment (mdboom) [reply src/preamble.js]: I'm having a little trouble turning what I know needs to happen into a patch.

Basically, the allocation of the static data area in a SIDE_MODULE is:

```
gb = alignMemory(getMemory(205120), 8 || 1);
```
```
/* memory initializer */ allocate([0,0,0,0,0,0,240,63...], "i8", ALLOC_NONE, gb);
```

where the amount of memory allocated is more than the memory explicitly initialized.  The additional memory at the end is supposed to be zero'd, but it isn't being zero'd, so it should just be a matter of writing a loop from the end of the explicitly memory initializer to the end of the static block.  However, these two lines are written in different parts of the code (`jsifier.js` and fastcomp's `JSBackend.cpp`) and I don't think either place has both pieces of information necessary to write this loop.  Any thoughts?

Comment (kripken) [reply src/preamble.js]: Hmm, how about doing the zeroing out right after the first line? You know `gb` and the size `205120` there. It's not maximally efficient I suppose, but dynamic linking will be overhauled eventually anyhow - can just leave a comment there about the overhead.

**Comment by mdboom** [comment]
I've updated it with your suggestion of clearing the data after allocation, but before the static initializer is written in.

File: src/jsifier.js [hunk 3]
```diff
@@ -432,7 +432,10 @@ function JSify(data, functionsOnly) {
...
+          // Ideally, for efficiency, this would only zero memory after the static initialized
```
Comment (kripken) [inline src/jsifier.js]: please also add a comment explaining why side modules need this zeroing out

File: src/jsifier.js [hunk 4]
```diff
@@ -432,7 +432,10 @@ function JSify(data, functionsOnly) {
...
+          // data, but we don't know the size of that here.
+          print('for (let i = gb; i < gb + {{{ STATIC_BUMP }}}; ++i) HEAP8[i] = 0;\n');
```
Comment (kripken) [inline src/jsifier.js]: while I agree `let` is nicer, please use `var` for consistency with the rest of the code. (might be issues with older VMs, too)

**Comment by kripken** [comment]
Adding a test for this in #6423, and while doing so I noticed this didn't yet handle alignment correctly, fixed there.

**Pull request merged**